src/turnkeyml/common/filesystem.py

```diff
@@ -528,3 +554,198 @@ def rebase_cache_dir(input_path: str, build_name: str, new_cache_dir: str):
+    initial arguments, through each build Stage in the Sequence, and finally
+    to the disk, where it is used to assess cache hits.
+
+    State is initialized with the key members that are shared by every build,
+    and reasonable default values are assigned as appropriate.
+
+    Stage developers can also add any members they wish. To get or set an
+    attribute, reference it as an attribute:
+        1. get: `my_variable = state.attribute_name`
+        1. set: `state.attribute_name = my_variable`
+
+    Build State can be saved and loaded from disk in the form of a state.yaml file
+    via State.save() and load_state(), respectively. Note that while State can
+    contain members of any type, only YAML-safe members (str, int, bool, float,
+    list, dict, tuple) will be saved and loaded.
+    """
+
+    def __init__(
+        self,
+        evaluation_id: str,
+        cache_dir: str,
+        model: Optional[build.UnionValidModelInstanceTypes] = None,
+        inputs: Optional[Dict[str, Any]] = None,
+        model_type: Optional[build.ModelType] = None,
+        monitor: Optional[bool] = None,
+        build_name: Optional[str] = None,
+        sequence=None,
+        onnx_opset: Optional[int] = None,
+        device: Optional[str] = None,
+        **kwargs,
+    ):
+
+        # Allow monitor to be globally disabled by an environment variable
+        if monitor is None:
+            if os.environ.get("TURNKEY_BUILD_MONITOR") == "False":
+                monitor_setting = False
+            else:
+                monitor_setting = True
+        else:
+            monitor_setting = monitor
+
+        # The default model name is the name of the python file that calls build_model()
+        if build_name is None:
+            build_name = os.path.basename(sys.argv[0])
+
+        if sequence is None:
+            # The value ["default"] indicates that build_model() will be assigning some
+            # default sequence later in the program
+            stage_names = ["default"]
+        elif isinstance(sequence, list):
+            stage_names = sequence
+        else:
+            stage_names = sequence.get_names()
+
+        # Detect and validate ONNX opset
+        if isinstance(model, str) and model.endswith(".onnx"):
+            onnx_file_opset = onnx_helpers.get_opset(onnx.load(model))
+
+            if onnx_opset is not None and onnx_opset != onnx_file_opset:
+                raise ValueError(
+                    "When using a '.onnx' file as input, the onnx_opset argument must "
+                    "be None or exactly match the ONNX opset of the '.onnx' file. However, the "
+                    f"'.onnx' file has opset {onnx_file_opset}, while onnx_opset was set "
+                    f"to {onnx_opset}"
+                )
+
+            opset_to_use = onnx_file_opset
+        else:
+            if onnx_opset is None:
+                opset_to_use = build.DEFAULT_ONNX_OPSET
```

Comment:
Do we want to upllift DEFAULT_ONNX_OPSET to 16 or 18 as a part of the refresh?

Reply:
That would be a good time to do this

Reply:
Opened https://github.com/onnx/turnkeyml/issues/164

Technically doesn't need to be part of the refresh since it isn't a massive breaking API change like everything else 😓 , could do it right after.